README

mars2mseed - convert MARS 88/lite waveform data to Mini-SEED.

All data samples are scaled by 10 during this conversion, do not forget
to take this into account when using the resulting Mini-SEED.

-- Acknowledgements --

Aladino Govoni provided most of the base code for reading and parsing
MARS data blocks and was very helpful, many thanks.


-- Building/Installing --

In most environments a simple 'make' will build the program.

Using GCC, running 'make static' will compile a static version
if possible.

For Win32 (Windows) Open Watcom's wmake program can be used with
Makefile.wat (e.g. wmake -f Makefile.wat), this has only been tested
with Open Watcom 1.3.  Some untested Makefile.win files exist that
might work with Microsoft's nmake program.

For usage information see the man page in the doc directory.

-- Licensing --

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License (GNU-GPL) for more details.  The GNU-GPL and
further information can be found here: http://www.gnu.org/



ChangeLog
2017.099: 1.4
	- Update libmseed to 2.19.3, adjust counters to 64-bit.
	- Remove dependency on ntwin32.mak for Windows nmake makefiles, now
	building works in plain MSVC development environments.
	- Add DEPRECATED notice to Makefile.wat and src/Makefile.wat.
	- Clean old targets out of src/Makefile.
	- Update licensing to GNU GPL 3.

2010.029: 1.3
	- Update to libmseed 2.4, including fixes for packing long
	segments of data.

2008.220: 1.2
	- Update libmseed to 2.1.6, including much improved Steim 1
	and 2 encoding performance.

2007.284: 1.1
	- Update to libmseed 2.1.4.
	- Change src/Makefile to use user defined CFLAGS when specified.

2006.363:
	- Update to libmseed 2.0rc1 + 2006.363.
	- Convert all output to ms_log().
	- Fix FPE when scaling is set to 0 (a useless value anyway).

2006.209: 
	- Remove -P option, packing after each block is now default.
	- Add -g option to specify a scaling factor and make the
	default scaling 8 which results in amplitude units of 125nV.
	- Add a warning message for when sample value truncation is
	occurring, only print once per input block.
	- Convert many long's to int's in MARS reading code to enable
	proper 64-bit compilation and operation.
	- Update to libmseed 1.8.  Includes memory leak fix.
	- Add gcc32, gcc32debug and gcc64debug targets to Makefile.
	
2006.201:
	- Correct message regarding output scaling, output is 100 nV.

2006.199:
	- Add -P option to pack data after each block is read in to
	minimize memory usage, allowing large file conversion.

2006.183: 1.0
	- Update to libmseed 1.7.

2005.300: 0.4
	- Update libmseed to 1.4.2.
	- Any input file prefixed with '@' is assumed to be a list file.
	- Default output file is now <inputfile>.mseed instead of stdout.
	- Pack data after reading each file.
	- Add option to buffer all data in memory before packing.
	- Remove signal handling, not needed.
	- Add man page.
	- Add Win32 Makefiles, test with Open Watcom 1.3.

2005.270: 0.3
	- Add EH[ZNE] preset mappings.	
	- Add custom channel number to channel codes mapping ability.
	- Add warning message that all samples have been scaled.
	- Update to libmseed 1.2.

2005.???: versions 0.1 and 0.2.


src/mars.h
/*
$Header: /usr/local/cvs/repository/mars2mseed/src/mars.h,v 1.3 2006-07-27 17:28:00 chad Exp $
*/
#ifndef MARS_H_
 #define MARS_H_

 /* typedef for TIMES structure */
 /* low nibble: Sync Mode. These are the same values used inside MARS-88! */
 #define SM_NOSYNC 0x01   /* Sync_mode: none             */
 #define SM_TSIG   0x02   /* Sync_mode: time signal      */
 #define SM_DCF77  0x04   /* Sync_mode: DCF              */

 /* hi  nibble: clock state. */
 #define DCF_OK    0x10   /* Time checked by DCF         */
 #define SYNC_OK   0x20   /* Clock synchronized          */

 #define TM_VALID  0x80

 #define NO_WORD   0x7fff /* leTime.delta=NO_WORD: no valid delta */
 
 #define LEm88Base	16
 #define LEliteBase	14

 #define M88_MINSAMPLING	0x07
 #define M88_MAXSAMPLING	0x01
 #define M88_CHANMASK		0x03

 /* m88 block status */
 #define m88HeadError		0x0FFF0000
 #define m88NoValidFormat	0x00020000
 #define m88NoValidDevID	0x00040000
 #define m88NoValidTime		0x00080000
 #define m88NoValidChan		0x00100000
 #define m88NoValidRate		0x00200000
 #define m88NoValidAmp		0x00400000
 #define m88NoValidScale	0x00800000

 #define m88DataError		0xF0000000
 #define m88MisplacedBlock	0x10000000
 #define m88TimeZappedBlock	0x20000000

 #define m88ResetHeadError(a)	( (a) = ( (a)&(~m88HeadError) ) )

 #define mlDiskBlockSize        512
 #define mlDiskPartInfo         (2*mlDiskBlockSize)
 #define mlPartInfoEntries      4

 #define marsBlockSamples       500
 #define marsBlockDataSize      1000
 #define marsBlockSize          1024

 typedef struct mlPartitionInfo
 {
  char	  label[8];
  int 	  offset,uoffset;
  int 	  length,ulength;
  int	  dummy,udummy;
 } mlPartInfo;
 
 #define mlDATA	      0
 #define mlMONITOR    1
 #define mlLOGGING    2
 #define mlSETUP      3
 
 #define LE_MAGIC	('l' + ('e'<<8))
 #define LEMAGICle	0x656C  /* 'l'+'e' value on little-endian */
 #define LEMAGICbe	0x6C65  /* 'l'+'e' value on big-endian    */
 
 #define DATABLK_FORMAT     1   /* MARS-88 data block 	  */
 #define MONBLK_FORMAT      2   /* MARS-88 monitor block  */
 #define LITE_BLOCK_FORMAT  3   /* MARSlite data block    */
 #define LITE_MONBLK_FORMAT 4   /* MARSlite monitor block */

 typedef struct
 {
  short	magic;		    /* 'l' 'e'         */
  char 	block_format;       /* block format ID */
  char 	data_format;        /* data format ID, same values as DSP   */
                            /* 0: straight 16-bit word              */
                            /* 1: 14 bits mantissa, 2 bits exponent, MARS-88 */
                            /* 2: 13 bits mantissa, 3 bits exponent, MARS-88 */
                            /* 3: 12 bits mantissa, 4 bits exponent, MARS-88 */
                            /* 4: 3 bits exponent/2, 13 bits mantissa, non-differential, lite */
                            /* 5: 3 bits exponent/2, 13 bits mantissa, differential, lite */
 } leFormat;

 typedef struct
 {
    int    time;	    /* time/date in ctime(3C) format */
    short  delta;	    /* time lag in ms                */
    short  mode;	    /* low nibble: sync mode; hi nibble: clock state */
 } leTime;

 typedef struct
 { 			    /*                         length  total   */
    leFormat format_id;	    /* header-data format ID      4      4     */
    int    dev_id;	    /* hardware device ID         4      8     */
    leTime time;	    /* time, date, time delta,..  8     16     */
    char   chno;	    /* channel number             1     17     */
    char   samp_rate;	    /* sampling rate (2^N ms)     1     18     */
    short  maxamp;	    /* max.amplitude              2     20     */
    char   scale;	    /* scale (2^N uV)             1     21     */
    char   dummy[3];	    /* ... padding                3     24     */
 } m88Head;

 typedef struct 
 {                          /*                          length  total   */
  leFormat format_id;       /* header-data format ID       4      4     */
  char     station_name[4]; /* device name                 4      8     */
  char     _rsrvd[4];       /* reserved                    4     12     */
  int      time;            /* time, date                  4     16     */
  char     chno;            /* channel number              1     17     */
  char     samp_interval;   /* sampling interval (2^N ms)  1     18     */
  short    maxamp;          /* max.amplitude               2     20     */
  char     scale;           /* scale (2^N uV)              1     21     */
  unsigned char triggidx;   /* trigger index / 2           1     22     */
  short    dstart;          /* start value f. diff format  2     24     */
 } mlHead;

 typedef struct
 {
    m88Head	head;
    short	data[marsBlockSamples];
 } m88Block;

 typedef struct
 {
    mlHead	head;
    short	data[marsBlockSamples];
 } mlBlock;

#endif


src/marsio.h
/*
$Header: /usr/local/cvs/repository/mars2mseed/src/marsio.h,v 1.4 2006-07-27 17:28:00 chad Exp $
*/
#ifndef MARSIO_H_
 #define MARSIO_H_

 #include <stdio.h>
 #include <time.h>
 
 #include <libmseed.h>

 #include "mars.h"
 
 #define msStreamActive       0x00000001 
 #define msLongHeaders        0x00000100
 
 #define msCheckStatus(a,b)   ( (a)&(b) )
 
 #define mbHeaderMacros
 #define mbGetMagic(a)	      ((((m88Head *)(a))->format_id).magic)
 #define mbGetBlockFormat(a)  ((((m88Head *)(a))->format_id).block_format)
 #define mbGetDataFormat(a)   ((((m88Head *)(a))->format_id).data_format)
 
 #define mbGetChan(a)	      (((m88Head *)(a))->chno)
 #define mbGetSamp(a)	      (1<<(((m88Head *)(a))->samp_rate))
 #define mbGetSampRate(a)     (1000.0/mbGetSamp(a))
 #define mbGetMaxamp(a)	      (((m88Head *)(a))->maxamp)
 #define mbGetScale(a)	      (1<<((m88Head *)(a))->scale)
 #define mbGetTime(a)	      ( ((((m88Head *)(a))->format_id).data_format < LITE_BLOCK_FORMAT) ? (((m88Head *)(a))->time).time : ((mlHead *)(a))->time  )

 typedef struct
 {
  FILE	*hf;
  off_t	offset;
  char	block[marsBlockSize];
  
  size_t  status;
  
  /*	file info	*/
  off_t		size;
  time_t	time;
  char 		name[4096];
 
 } marsStream;
 
/*********************************************************
***   Function Prototypes
**********************************************************/
 #ifdef __cplusplus
  extern "C" {
 #endif
 
 marsStream *marsStreamOpen(char *name);
 marsStream *marsStreamGetNextBlock(int verbose);
 marsStream *marsStreamGetCurrent(void);
 
 void marsStreamClose(void);
 void m88SwapBlock(m88Block *blk);
 
 int *marsBlockDecodeData(char *block,int *scale);
 int marsStreamDumpBlock(marsStream *hMS);

 double marsBlockGetGain(char *blk);
 int marsBlockGetScaleFactor(char *blk);
 char *mbGetStationCode(char *blk);
 int mbGetStationSerial(char *blk);
 
 #ifdef __cplusplus
  }
 #endif
#endif


src/marsio.c
/*
Revision 1.8  2006-12-29 11:00:00-08  ctrabant
Change gswapX to ms_gswapX and use ms_log()

Revision 1.7  2006-07-27 16:55:00-08  ctrabant
Change long's to int's to make 64-bit possible

Revision 1.6  2005-05-03 14:00:00-08  ctrabant

Revision 1.5  2005-02-17 14:00:00-08  ctrabant
Modifications for mars2mseed.

Revision 1.4  2005-02-04 14:31:32+00  aladino
Backup revision

Revision 1.3  2005-02-04 13:11:49+00  aladino
Backup revision

Revision 1.2  2005-02-03 16:24:04+00  aladino
Backup revision

Revision 1.1  2005-02-03 14:27:13+00  aladino
Backup revision
*/

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#include <sys/types.h> 
#include <sys/stat.h> 

#include <string.h>
#include <math.h>

#include <libmseed.h>

#include "marsio.h"

static marsStream MS;
static int m88BlockDecodedData[marsBlockSamples];
static char mbNameBuf[64];


int isMarsDataBlock (char *blk)
{
  m88Head *head = (m88Head *)blk;
  
  return ( (((head->format_id).magic == LEMAGICle) || ((head->format_id).magic == LEMAGICbe)) &&
	   (((head->format_id).block_format == DATABLK_FORMAT) ||
	    ((head->format_id).block_format == LITE_BLOCK_FORMAT)) );
}


char *mbGetStationCode (char *blk)
{
  switch( mbGetBlockFormat(blk) )
    {
    case DATABLK_FORMAT:
      sprintf (mbNameBuf, "%04X", (unsigned int)(((m88Head *)blk)->dev_id & 0xFFFF));
      break;
    case LITE_BLOCK_FORMAT:
      strcpy (mbNameBuf, ((mlHead*)blk)->station_name);
      break;
    default:
      sprintf (mbNameBuf, "????");
      break;
    }
  
  return mbNameBuf;
}


int mbGetStationSerial (char *blk)
{
  switch( mbGetBlockFormat(blk) )
    {
    case DATABLK_FORMAT:
      return (int)(((m88Head *)blk)->dev_id & 0xFFFF);
      break;
    case LITE_BLOCK_FORMAT:
      return 0;
      break;
    default:
      return -1;
      break;
    }
  
  return -1;
}


int convert_word (short input, short format)
{
  int exponent;
  int mantissa;
  int shift;
  
  switch (format)
    {
    case 0:
      return (int) input;
      break;
    case 1:
      mantissa = input & (~0x03);
      exponent = input & 0x03;
      return ((int) mantissa << (16 - exponent));
      break;
    case 2:
      mantissa = input & (~0x07);
      exponent = input & 0x07;
      return ((int) mantissa << (16 - exponent));
      break;
    case 3:
      mantissa = input & (~0x0f);
      exponent = input & 0x0f;
      return ((int) mantissa << (16 - exponent));
      break;
    /***** MARSlite data format here! ************************/
    case 4:    /* MARSlite format; non-differential and     */
    case 5:    /* differential use the same word format     */
      mantissa = input & (~0x07);
      exponent = input & 0x07;
      shift = 2 * exponent;
      return ((int) mantissa << (16 -shift));
      /* your application program will want to convert this to a      */
      /* floating point number, taking the "scale" value into account */
      break;
    default:
      ms_log (2, "convert_word(): illegal data format %d\n", format);
      return 0L;
    }
}


int marsBlockGetScaleFactor (char *blk)
{
/*	Negative values means multiply
	Positive means divide
	0 invalid
*/
  m88Head   *head=(m88Head *)blk;
  int 	     scale;
  
  switch ( mbGetDataFormat(head) )
    {
    case 0:
      scale = -(1 << head->scale);
      break;
    case 1:
    case 2:
    case 3:
      scale = (1<<(LEm88Base - head->scale));
      break;
    case 4:
    case 5:
      scale = (1<<(LEliteBase - head->scale));
      break;
    default: /* invalid data format */
      scale = 0;
      ms_log (2, "marsBlockGetScaleFactor() : invalid data format 0x%1X\n",
	      mbGetDataFormat(head));
    }
  
  return scale;
}


double marsBlockGetGain (char *blk)
{
  m88Head   *head=(m88Head *)blk;
  double     gain=0;
  
  switch ( mbGetDataFormat(head) )
    {
    case 0:
      gain = (double)( 1 << head->scale );
      break;
    case 1:
    case 2:
    case 3:
      gain = 1.0 / (double)(1 << (LEm88Base - head->scale));
      break;
    case 4:
    case 5:
      gain = 1.0 / (double)(1 << (LEliteBase - head->scale));
      break;
    default:
      gain = 0;
      ms_log (2, "marsBlockGetGain() : invalid data format 0x%1X\n",
	      (head->format_id).data_format);
    }
  
  return gain;
}


int *marsBlockDecodeData (char *block, int *scale)
{
  int *data=m88BlockDecodedData;
  m88Block *buf=(m88Block *)block;
  
  int    i;
  int    data_format;
  int   mantissa;
  short  temp;
  int    exponent;
  int	 block_duration;
  int	 samp_interval;
  int    shift;
  int	 sum = 0;
  short  codedsum;

  /*
   *	correct for timing error in MARS-88 data for sampling intervals >= 32 msec.
   *	In these cases, blocks are delayed by one block length. The block
   *	length can be calculated: 1 << (samp -1 ) 
   *	where samp is the 2's log of the sampling interval.
   */
  samp_interval = buf->head.samp_rate;
  if (samp_interval >= 5)
    {
      block_duration =  1 << (samp_interval-1);
      buf->head.time.time -= block_duration;
    }
  
  if (buf->head.format_id.block_format == MONBLK_FORMAT)
    {
      data_format = 0;
    }
  else
    {
      data_format = buf->head.format_id.data_format;
    }
  if (data_format == 5)
    {
      codedsum=((mlHead *)block)->dstart;
      sum = convert_word (codedsum, data_format);
      /* P.L.Bragato:	QUESTO per il vecchio formato di uscita di "lt2m88"
	 Chad's translation: THIS for the old format of escape of "lt2m88"
	 codedsum = ((buf->head.dummy[0]) << 8) | ( (buf->head.dummy[1]) & 0x00ff);
	 sum = convert_word (codedsum, data_format);
      */
#ifdef DEBUG
      ms_log (1, "START at %d (0x%04x\n",sum,(buf->head.dstart)&0xffff);
#endif
    }	
  
  switch (data_format) {
  case 0:
    for (i = 0; i < 500; i++)
      {
	data[i] = (int) buf->data[i];
      }
    *scale=(1<<(buf->head).scale);
    break;
  case 1:
    for (i = 0; i < 500; i++)
      {
	temp = buf->data[i];
	mantissa = temp & (~0x03);
	exponent = temp & 0x03;
	data[i] = (int) mantissa << (16 - exponent);
      }
    *scale=(1<<(LEm88Base -(buf->head).scale));
    break;
  case 2:
    for (i = 0; i < 500; i++)
      {
	temp = buf->data[i];
	mantissa = temp & (~0x07);
	exponent = temp & 0x07;
	data[i] = (int) mantissa << (16 - exponent);
      }
    *scale=(1<<(LEm88Base -(buf->head).scale));
    break;
  case 3:
    for (i = 0; i < 500; i++)
      {
	temp = buf->data[i];
	mantissa = temp & (~0x0f);
	exponent = temp & 0x0f;
	data[i] = (int) mantissa << (16 - exponent);
      }
    *scale=(1<<(LEm88Base -(buf->head).scale));
    break;
  /***** MARSlite data format here! ************************/
  case 4:    /* MARSlite format; non-differential   */
    for (i=0; i<500; i++)
      {
	temp = buf->data[i];
	mantissa = temp & (~0x07);
	exponent = temp & 0x07;
	shift = 2 * exponent;
	data[i] = (int) mantissa << (16 - shift);
      } 
    *scale=(1<<(LEliteBase -(buf->head).scale));
    break;
  case 5:    /* differential */
    for(i=0; i<500; i++)
      {
	temp = buf->data[i];
	exponent = temp & 0x07;
	mantissa = temp & (~0x07);
	shift = 2 * exponent;
	data[i] = (int) mantissa << (16 - shift);
	sum += data[i];
#ifdef DEBUG
	ms_log (1, "%3d: delta %12ld (0x%04x mant 0x%4x exp 0x%1x) sum %12ld\n",
		i,data[i],(buf->data[i])&0xffff,mantissa&0xffff,exponent&0xf,sum);
#endif
	data[i] = sum;
      }
    *scale=(1<<(LEliteBase -(buf->head).scale));
    break;
    
  default:
    ms_log (2, "marsBlockDecodeData(): illegal data format %d\n", data_format);
    return NULL;
  }

  return data;
}


void m88SwapBlock (m88Block *block)
{
  m88Head  *head = &block->head;
  short    *data;
  
  /* Swap header */
  ms_gswap4 (&head->dev_id);
  ms_gswap4 (&(head->time).time);
  ms_gswap2 (&(head->time).delta);
  ms_gswap2 (&(head->time).mode);
  ms_gswap2 (&head->maxamp);
  
  /* Swap data */
  for ( data = block->data; data < block->data+marsBlockSamples; data++)
    ms_gswap2 (data);
}


void mlSwapBlock (mlBlock *block)
{
  mlHead  *head = &block->head;
  short   *data;
  
  /* Swap header */
  ms_gswap4 (&head->time);
  ms_gswap2 (&head->maxamp);
  ms_gswap2 (&head->dstart);
  
  /* Swap data */
  for ( data = block->data; data < block->data+marsBlockSamples; data++)
    ms_gswap2 (data);
}


marsStream *marsStreamOpen (char *name)
{
  struct stat	fs;
  
  memset (&MS, 0, sizeof(marsStream));
  
  if ( stat(name,&fs) )
    {
      ms_log (2, "Cannot stat file \'%s\' - %s\n", name, strerror(errno));
      return NULL;
    }
  
  if ( (MS.hf = fopen(name,"rb")) == NULL )
    {
      ms_log (2, "Cannot open file \'%s\' - %s\n", name, strerror(errno));
      return NULL;
    }
  
  /* fill marsStream structure */
  MS.size=fs.st_size;
  MS.time=fs.st_mtime;
  strcpy (MS.name, name);
  
  MS.status|=msStreamActive;
  
  return &MS;
}


int marsStreamDumpBlock (marsStream *hMS)
{
  char 	*hB=hMS->block;
  char   timestr[50];
  hptime_t hptime;
  int	*hData, *hD, scale;
  double gain;
  
  if ( isMarsDataBlock(hB) )
    {
      hData = marsBlockDecodeData (hB,&scale);
      hptime = MS_EPOCH2HPTIME (mbGetTime(hB));
      ms_hptime2isotimestr (hptime, timestr, 1);
      ms_log (1, "MB sta='%4s' chano=%d block=%d samp=%d scale=%d time=%s c2uV=%d maxamp=%d",
	      mbGetStationCode(hB), mbGetChan(hB), mbGetBlockFormat(hB),
	      mbGetSamp(hB), mbGetScale(hB),
	      timestr,
	      marsBlockGetScaleFactor(hB),
	      mbGetMaxamp(hB));
      
      if ( hData )
	{
	  ms_log (1, "\n");
	  for (hD=hData, gain=marsBlockGetGain(hB); hD<(hData+marsBlockSamples); hD++)
	    ms_log (1, "%d => %g\n",*hD, *hD*gain);
	  return 1;
	}
      else
	ms_log (2, ":ERROR: CANNOT DECODE\n");
    }
  
  return 0;
}


marsStream *marsStreamGetNextBlock (int verbose)
{
  
  while ( fread(MS.block, marsBlockSize, 1, MS.hf) == 1 )
    {
      /* Byte swap block if necessary (i.e. host is big-endian) */
      if ( mbGetMagic(MS.block) == LEMAGICbe )
	switch ( ((leFormat *) &(MS.block))->block_format )
	  {
	  case 1:  /* MARS-88 data block */
	  case 2:  /* MARS-88 monitor block */
	    m88SwapBlock ((m88Block *) &(MS.block));
	    break;
	  case 3:  /* MARSlite data block */
	  case 4:  /* MARSlite monitor block */
	    mlSwapBlock ((mlBlock *) &(MS.block));
	    break;
	  }
      
      if ( verbose >= 2 )
	ms_log (1, "MB 0x%016X : block %d : %s : 0x%04X : %d : %d : chan %d\n",
		(unsigned int)MS.offset,(int)(MS.offset/marsBlockSize),
		isMarsDataBlock(MS.block)?"DATA":"MON ",
		mbGetMagic(MS.block),mbGetBlockFormat(MS.block),mbGetDataFormat(MS.block),
		mbGetChan(MS.block));
      
      if ( isMarsDataBlock(MS.block) && mbGetChan(MS.block) < 3 )
	{ /* do checks */
	  MS.offset += marsBlockSize;
	  return &MS;
	}
      
      MS.offset += marsBlockSize;
    }
  
  return NULL;
}


marsStream *marsStreamGetCurrent (void)
{
  return &MS;
}


void marsStreamClose (void)
{
  if ( MS.hf != NULL )
    fclose (MS.hf);
  
  memset (&MS, 0, sizeof(marsStream));
}
